ts/nni_manager/common/trial_keeper/keeper.ts

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ *  The helper class to implement a "reuse" training service.
+ *
+ *  TrialKeeper has a very similar interface to TrainingSerivceV3.
+ *  In fact the local training service is a thin wrapper over TrialKeeper.
+ **/
+
+import { EventEmitter } from 'events';
+import fs from 'fs/promises';
+import path from 'path';
+
+import type { Command } from 'common/command_channel/interface';
+import { HttpChannelServer } from 'common/command_channel/http';
+import globals from 'common/globals';
+import { Logger, getLogger } from 'common/log';
+import { TrialProcess, TrialProcessOptions } from './process';
+import { GpuSchedulerClient } from './gpu_scheduler_client';
+
+export declare namespace TrialKeeper {
+    export interface TrialOptions {
+        id: string;
+        command: string;
+        codeDirectoryName: string;
+        sequenceId?: number;
+        gpuNumber?: number;
+        gpuRestrictions?: GpuRestrictions;
+    }
+
+    export interface GpuRestrictions {
+        onlyUseIndices?: number[];
+        rejectActive?: boolean;
+        rejectComputeActive?: boolean;
+    }
+}
+
+export class TrialKeeper {
+    private envId: string;
+    private channels: HttpChannelServer;
+    private dirs: Map<string, string> = new Map();
+    private emitter: EventEmitter = new EventEmitter();
+    private gpuScheduler: GpuSchedulerClient;
```

Comment:
I'm still feeling a little uncomfortable with `gpuScheduler`. Maybe `scheduler`?

Reply:
There will also be a scheduler between NNI manager & training service, what should be its name?

Reply:
Maybe it's `taskScheduler`. The other one is `trialScheduler`.

As long as the interface is clear, I think name is the easy one.

Reply:
how about "trialScheduler" and "resourceAllocator"?